src/workers/helpers.js

```diff
@@ -1,53 +1,238 @@
+    }
+  }
+
+  return output;
+}
+
+export async function fetchStepSummary(state, step) {
+  // do not send any response to UI if they have not changed
+  if (!state[step].changed) {
+    return null;
+  }
+
+  if (step === "inputs") {
+    let output = {};
+
+    let ngenes = {};
+    for (const a of state[step].fetchCountMatrix().available()) {
+      ngenes[a] = state[step].fetchCountMatrix().get(a).numberOfRows();
+    }
+
+    let gene_info = {};
+    for (const [k, v] of Object.entries(
+      state[step].fetchFeatureAnnotations()
+    )) {
+      let info = {};
+      for (const c of v.columnNames()) {
+        info[c] = bioc.CLONE(v.column(c));
+      }
+      gene_info[k] = info;
+    }
+
+    output = {
+      num_cells: state[step].fetchCountMatrix().numberOfColumns(),
+      num_genes: ngenes,
+      genes: gene_info,
+      annotations: state[step].fetchCellAnnotations().columnNames(),
+    };
+
+    return output;
+  } else if (step === "rna_quality_control") {
+    let output = {};
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    return output;
+  } else if (step === "adt_quality_control") {
+    var output = {};
+
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    // We don't use sums for filtering but we do report it in the metrics,
+    // so we just add some NaNs to the thresholds for consistency.
+    for (const [k, v] of Object.entries(output.thresholds)) {
+      v.sums = NaN;
+    }
+
+    return output;
+  } else if (step === "crispr_quality_control") {
+    let output = {};
+    var blocks = state["inputs"].fetchBlockLevels();
+    if (blocks === null) {
+      blocks = ["default"];
+      output.data = { default: state[step].fetchMetrics() };
+    } else {
+      let metrics = state[step].fetchMetrics({ copy: "view" });
+      let bids = state["inputs"].fetchBlock();
+      output.data = splitMetricsByBlock(metrics, blocks, bids);
+    }
+
+    let listed = state[step].fetchFilters();
+    output.thresholds = splitThresholdsByBlock(listed, blocks);
+
+    return output;
+  } else if (step === "cell_filtering") {
```

Comment:
This entire step could be simplified by just checking `state[step].fetchFilteredMatrix().numberOfColumns()`, which is the number retained. If you need the number lost (don't see why), you can subtract the previous value from `state.inputs.fetchCountMatrix().numberOfColumns()`.